app/services/data_exports.rb

```diff
@@ -0,0 +1,94 @@
+require 'fileutils'
+require 'securerandom'
+
+module DataExports
+
+  class DataExporter
+    def export_transcription(transcription_id)
+      file_generator = TranscriptionFileGenerator.new(transcription_id)
+      file_list = file_generator.generate_transcription_files
+
+      azure = AzureBlobStorage.new
+      azure.put_files_multiple(file_list)
+
+      # we are done with the files, delete the temp directory and its contents
+      file_generator.delete_temp_directory
+    end
+  end
+
+  class TranscriptionFileGenerator
+    def initialize(transcription_id)
+      @transcription = Transcription.find(transcription_id)
+      @directory_path = File.expand_path("~/transcription_files_temp/t#{transcription_id}_#{SecureRandom.uuid}")
+
+      FileUtils.mkdir_p @directory_path
+    end
+
+    def generate_transcription_files
+      file_list = []
+      blob_directory = generate_blob_directory(@transcription)
+
+      # raw data file
+      file_path = write_raw_data_to_file(@directory_path)
+      blob_path = File.join(blob_directory, "raw_data_#{@transcription.id}.json")
+      file_list.append({ :blob_path => blob_path, :file => file_path})
+
+      # consensus text file
+      file_path = write_consensus_text_to_file(@directory_path)
+      blob_path = File.join(blob_directory, "consensus_text_#{@transcription.id}.txt")
+      file_list.append({ :blob_path => blob_path, :file => file_path})
+    end
+
+    def delete_temp_directory
+      FileUtils.remove_dir(@directory_path)
+    end
```

Comment:
Ideally, this method would get run automatically upon the instance of the `TranscriptionFileGenerator` going out of scope. Does anyone have ideas on how to do this?

Reply:
Nope - you're really talking about destructors in ruby (https://stackoverflow.com/questions/5956067/ruby-destructors)

Best way is to use ruby temp files as they have this behaviour built in or look at using tmp dirs, https://ruby-doc.org/stdlib-2.5.3/libdoc/tmpdir/rdoc/Dir.html

Reply:
When I was first setting up the connection to azure, passing in ruby tempfiles was giving me issues (the tempfile seemed to behave differently than a normal file, I wasn't able to get the actual content of the file to save to azure) - I will give it another go, with Active Storage it might not even give me any issues. I will let you know how it goes!

Reply:
Happy to help / pair on it if you need.

Reply:
Turns out that using tempfiles for the upload process was straightforward. But as for switching to using tempfiles when downloading and zipping files, it's a little trickier. I'd love to talk a bit about this with you, here's my current thoughts on how I'd go about it:
* Let top-level directory be a tempdir
* Create normal directories and files within that tempdir because we need to be able to specify names for all dirs and files that'll be passed to the zip file generator
* Should also be possible to have the generated zip file be a tempfile

I'll dm you on slack, would be nice to talk it over with you.